server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/IdentityDAO.scala

```diff
@@ -0,0 +1,113 @@
+}
+case class IdentityHtmlSignatureUpdate(htmlSignature: Option[HtmlSignature]) extends IdentityUpdate {
+  override def update(identity: Identity): Identity = identity.copy(htmlSignature = htmlSignature)
+}
+
+case class IdentityUpdateRequest(name: IdentityNameUpdate,
+                                 replyTo: IdentityReplyToUpdate,
+                                 bcc: IdentityBccUpdate,
+                                 textSignature: IdentityTextSignatureUpdate,
+                                 htmlSignature: IdentityHtmlSignatureUpdate) {
+  def update(identity: Identity): Identity =
+    List(name, replyTo, bcc, textSignature, htmlSignature)
+      .foldLeft(identity)((acc, update) => update.update(acc))
+}
+
+trait CustomIdentityDAO {
+  def save(user: Username, creationRequest: IdentityCreationRequest): Publisher[Identity]
+
+  def list(user: Username): Publisher[Identity]
+
+  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit]
+
+  def delete(username: Username, ids: Seq[IdentityId]): Publisher[Unit]
+}
+
+class DefaultIdentitySupplier @Inject()(canSendFrom: CanSendFrom) {
+  def listIdentities(username: Username): List[Identity] =
+    canSendFrom.allValidFromAddressesForUser(username)
+      .collect(ImmutableList.toImmutableList()).asScala.toList
+      .flatMap(address =>
+        from(address).map(id =>
+          Identity(
+            id = id,
+            name = IdentityName(address.asString()),
+            email = address,
+            replyTo = None,
+            bcc = None,
+            textSignature = None,
+            htmlSignature = None,
+            mayDelete = MayDeleteIdentity(false))))
+
+  private def from(address: MailAddress): Option[IdentityId] =
+    Try(UUID.nameUUIDFromBytes(address.asString().getBytes(StandardCharsets.UTF_8)))
+      .toEither
+      .toOption
+      .map(IdentityId)
+}
+
+// This class is intended to merge default (server-set0 identities with (user defined) custom identities
+// Using the custom identities we can stores deltas of the default (server-set) identities allowing to modify them.
+class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identityFactory: DefaultIdentitySupplier) {
+  def save(user: Username, creationRequest: IdentityCreationRequest): Publisher[Identity] = customIdentityDao.save(user, creationRequest)
+
+  def list(user: Username): Publisher[Identity] = SFlux.merge(Seq(
+    customIdentityDao.list(user),
+    SMono.fromCallable(() => identityFactory.listIdentities(user))
+      .subscribeOn(Schedulers.elastic())
+      .flatMapMany(SFlux.fromIterable)))
+
+  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
+
+  def delete(username: Username, ids: Seq[IdentityId]): Publisher[Unit] = customIdentityDao.delete(username, ids)
```

Comment:
Should we check Id is exists in `identityFactory`, if yes => throw an exception?

From https://jmap.io/spec-mail.html#identityset
```
Servers may wish to set this to false for the user’s username or other default address. Attempts to destroy an Identity with mayDelete: false will be rejected with a standard forbidden SetError.
```

Reply:
I let later tickets do this ;-)